### DISCUSSION TITLE
Errors when trying to compile "hello world" app with MCUBOOT enabled for pinetime_devkit0 board

### DISCUSSION THREAD

**Discussion by ljunquera** [zephyrproject-rtos/zephyr#62476] [Q&A]
I am getting errors when trying to compile "hello world" app with MCUBOOT enabled for [pinetime_devkit0](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/pinetime_devkit0/pinetime_devkit0.dts) board. I am able to build with [nrf52dk_nrf52832](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/nrf52dk_nrf52832/nrf52dk_nrf52832.dts). The errors are:

- ld.bfd: zephyr/drivers/spi/libdrivers__spi.a(spi_nrfx_spi.c.obj): in function `k_sem_give':
- [[undefined reference to `z_impl_k_sem_give'](kernel.h:1089: undefined reference to `z_impl_k_sem_give')](kernel.h:1089: undefined reference to `z_impl_k_sem_give')
- kernel.h:1089: undefined reference to `z_impl_k_sem_give'
- ld.bfd: zephyr/drivers/spi/libdrivers__spi.a(spi_nrfx_spi.c.obj): in function `k_sem_take':
- kernel.h:1072: undefined reference to `z_impl_k_sem_take'
- libdrivers__spi.a(spi_nrfx_spi.c.obj): in function `k_sem_give':
- kernel.h:1089: undefined reference to `z_impl_k_sem_give'
- libdrivers__spi.a(spi_nrfx_spi.c.obj): in function `k_sem_take':
- kernel.h:1072: undefined reference to `z_impl_k_sem_take'
- collect2: error: ld returned 1 exit status
- FAILED: modules/mcuboot/mcuboot_subimage-prefix/src/mcuboot_subimage-stamp/mcuboot_subimage-build mcuboot/zephyr/zephyr.hex mcuboot/zephyr/zephyr.elf 

I read [this on MCUboot](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.zephyrproject.org%2Flatest%2Fservices%2Fdevice_mgmt%2Fdfu.html%23mcuboot&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723838357%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=X9cgP2JISbqiLI2hQw065Oyjwl9wiu%2BNMqksU08sulc%3D&reserved=0) which outlines this:

1. You will need to define the flash partitions required by MCUboot; see [Flash map](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.zephyrproject.org%2Flatest%2Fservices%2Fstorage%2Fflash_map%2Fflash_map.html%23flash-map-api&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=0b%2FBq53YeW3g7HAgfsYd1%2FvxqUm1GhC%2FhjpPSJRhmDE%3D&reserved=0) for details.
2. You will have to specify your flash partition as the chosen code partition
**/ {   chosen {      zephyr,code-partition = &slot0_partition;   };};**

3. Your application’s .conf file needs to enable the [CONFIG_BOOTLOADER_MCUBOOT](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.zephyrproject.org%2Flatest%2Fkconfig.html%23CONFIG_BOOTLOADER_MCUBOOT&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=lpvNijUSg5sV58oD5G7zqv3E%2BmM6GQ7RvvvSCY%2BtxWQ%3D&reserved=0) Kconfig option in order for Zephyr to be built in an MCUboot-compatible manner

4. You need to build and flash MCUboot itself on your device

5. You might need to take precautions to avoid mass erasing the flash and also to flash the Zephyr application image at the correct offset (right after the bootloader)

I believe the [pinetime_devkit0 board definition](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/pinetime_devkit0/pinetime_devkit0.dts) meets the requirements and that my prj.conf is configured properly (it's working with nrf52dk_nrf52832):
1. I see this in the pinetime_devkit0.dts and I see the partitions:
- [boot_partition](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fblob%2Fbc695c6df5eb8a1f8f9720b20d7e5dcd1533242b%2Fboards%2Farm%2Fpinetime_devkit0%2Fpinetime_devkit0.dts%23L210C15-L210C15&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=dDEDqOxSXc1U3ChXOVP78Alsr1z2nDshEn2g8YHo%2FUw%3D&reserved=0)
```
		/* MCUboot bootloader */
		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x00000000 0xc000>;
		};
```
- [slot0_partition](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fblob%2Fbc695c6df5eb8a1f8f9720b20d7e5dcd1533242b%2Fboards%2Farm%2Fpinetime_devkit0%2Fpinetime_devkit0.dts%23L216C7-L216C7&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=v5yM%2BqRYub0rSmFo78mA%2BDeVqbDkpFTnMBJDMpajicc%3D&reserved=0)
```
		/* main firmware partition */
		slot0_partition: partition@c000 {
			label = "image-0";
			reg = <0x0000C000 0x74000>;
		};
```
- [slot1_partition](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fblob%2Fbc695c6df5eb8a1f8f9720b20d7e5dcd1533242b%2Fboards%2Farm%2Fpinetime_devkit0%2Fpinetime_devkit0.dts%23L161&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=XYFPwsS29riXrWNVAphJFiao4IfWdnUSCX4XCwEewBo%3D&reserved=0)
```
			/* next firmware update */
			slot1_partition: partition@0 {
				label = "image-1";
				reg = <0x00000000 0x74000>;
			};
```
2. I also see the [chosen code partition](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fblob%2Fbc695c6df5eb8a1f8f9720b20d7e5dcd1533242b%2Fboards%2Farm%2Fpinetime_devkit0%2Fpinetime_devkit0.dts%23L27&data=05%7C01%7C%7Cd4a68c294a6c4d94d42a08dbb20b0a8f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638299532723994569%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=mtzHuZj2jGH8x3s%2Bo43VSeV8u%2FjmzA1Hu9FCWYGnwvM%3D&reserved=0).
3. I have the CONFIG_BOOTLOADER_MCUBOOT=y in the prj.conf
4. I have built this code with a different board (nrf52832 devkit) with success, so I know I can build
5. I'm not even trying to flash at the moment.

I'm new to this and don't fully understand the errors. I'm assuming this is a configuration issue and may need an overlay file or something, but I'm not sure what to put in it if everything already in the pinetime_devkit0 definition.

**Comment by JacquesWIII** [comment]
Hey did you ever find a fix for this ?

**Reply by ljunquera** [reply]
I did, I just don't remember what I did. Probably some change to prj.conf.
